c10/macros/Macros.h

```diff
@@ -36,18 +36,32 @@
 
 /// C10_NODISCARD - Warn if a type or return value is discarded.
 #define C10_NODISCARD
-#if __cplusplus > 201402L && defined(__has_cpp_attribute)
-#if __has_cpp_attribute(nodiscard)
-#undef C10_NODISCARD
-#define C10_NODISCARD [[nodiscard]]
-#endif
+#if defined(__CUDACC__)
+// Unconditionally use [[nodiscard]] when compiling with nvcc; it
+// works (nvcc seems to support it, even though we're technically
+// not modern enough C++20); in any case, clang::warn_unused_result
+// doesn't work; see https://github.com/pytorch/pytorch/issues/13118
+#  undef C10_NODISCARD
+#  define C10_NODISCARD [[nodiscard]]
+// This code is dead (as we compile C++11 at the moment), but it is
+// the aspirationally correct implementation of this macro.
+// NB: You could make this trigger more frequently by dropping the C++
+// version check; some compilers (like nvcc) advertise C++14, but
+// __has_cpp_attribute(nodiscard) reports true anyway.
+#elif __cplusplus > 201402L && defined(__has_cpp_attribute)
+# if __has_cpp_attribute(nodiscard)
+#  undef C10_NODISCARD
+#  define C10_NODISCARD [[nodiscard]]
+# endif
 // Workaround for llvm.org/PR23435, since clang 3.6 and below emit a spurious
 // error when __has_cpp_attribute is given a scoped attribute in C mode.
 #elif __cplusplus && defined(__has_cpp_attribute)
-#if __has_cpp_attribute(clang::warn_unused_result)
-#undef C10_NODISCARD
-#define C10_NODISCARD [[clang::warn_unused_result]]
-#endif
+# if __has_cpp_attribute(clang::warn_unused_result)
+#  ifndef __CUDACC__
```

Comment:
Is this `if` necessary? Isn't it always true that `__CUDACC__ ` is not defined when entering this branch? Or maybe I don't understand what `defined(__CUDACC__)` does.

Reply:
No, it's not necessary. I'll remove it.